### PR TITLE
Update handling of betainc edge cases

### LINKED ISSUE #21900: Edge behavior in `jax.scipy.special.betainc`
### Description

`jax.scipy.special.betainc` seems to have trouble with very small values of the parameter `a`, at least for certain values of `b` and `x`.

```python3
import matplotlib.pyplot as plt
import numpy as np
from scipy.special import betainc as betainc_scipy
import jax.numpy as xp
from jax.scipy.special import betainc as betainc_jax

a = np.logspace(-40, -1, 300)
b = 1
x = 0.25
plt.loglog(a, betainc_scipy(a, b, x), label='scipy')
plt.loglog(a, betainc_jax(xp.asarray(a), b, x), label='jax')
plt.xlabel('a')
plt.ylabel('betainc(a, 1, 0.25)')
plt.legend()
```
image

I know that it is difficult to guarantee accuracy to machine precision for all possible combinations of input : ) Just thought I'd point out this problem spot since it came up in SciPy testing (scipy/scipy#20963).

---

There is a separate issue related to edge case behavior in `betainc` that I thought I should bring up: there are some edge cases where `betainc` returns NaN but another result would be more useful (e.g. in Harrell-Davis quantile estimates). To some extent, the "correct" result is a matter of interpretation, but SciPy recently addressed similar cases in scipy/scipy#22425. While we're waiting for an Array API special function extension (data-apis/array-api#725), please consider adopting this behavior.

<details>

```py
import jax.numpy as np
from jax.scipy import special
cases = [((0.0, 0.0, 0.0), np.nan),
         ((0.0, 0.0, 0.5), np.nan),
         ((0.0, 0.0, 1.0), np.nan),
         ((np.inf, np.inf, 0.0), np.nan),
         ((np.inf, np.inf, 0.5), np.nan),
         ((np.inf, np.inf, 1.0), np.nan),
         ((0.0, 1.0, 0.0), 0.0),
         ((0.0, 1.0, 0.5), 1.0),
         ((0.0, 1.0, 1.0), 1.0),
         ((1.0, 0.0, 0.0), 0.0),
         ((1.0, 0.0, 0.5), 0.0),
         ((1.0, 0.0, 1.0), 1.0),
         ((0.0, np.inf, 0.0), 0.0),
         ((0.0, np.inf, 0.5), 1.0),
         ((0.0, np.inf, 1.0), 1.0),
         ((np.inf, 0.0, 0.0), 0.0),
         ((np.inf, 0.0, 0.5), 0.0),
         ((np.inf, 0.0, 1.0), 1.0),
         ((1.0, np.inf, 0.0), 0.0),
         ((1.0, np.inf, 0.5), 1.0),
         ((1.0, np.inf, 1.0), 1.0),
         ((np.inf, 1.0, 0.0), 0.0),
         ((np.inf, 1.0, 0.5), 0.0),
         ((np.inf, 1.0, 1.0), 1.0)]

for args, ref in cases:
    print(special.betainc(*args), ref)
```

Produces: 
(actual, desired)
```
nan nan
nan nan
nan nan
nan nan
nan nan
nan nan
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
nan 0.0
nan 1.0
nan 1.0
nan 0.0
nan 0.0
nan 1.0
```

</details>

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.25.2
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='e901fac133dc', release='6.1.85+', version='#1 SMP PREEMPT_DYNAMIC Sun Apr 28 14:29:16 UTC 2024', machine='x86_64')

### PR DESCRIPTION
**Pull request by pearu**
As in the title, per https://github.com/scipy/scipy/pull/22425.

Fixes https://github.com/jax-ml/jax/issues/21900

### CODE REVIEW

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/test_util.py [hunk 1]
```diff
@@ -1518,7 +1518,11 @@ def _CheckAgainstNumpy(self, numpy_reference_op, lax_op, args_maker,
     args = args_maker()
     lax_ans = lax_op(*args)
     numpy_ans = numpy_reference_op(*args)
-    self.assertAllClose(numpy_ans, lax_ans, check_dtypes=check_dtypes,
+    # assertAllClose(x, y) expands to calling
```
[the remark this answers is not in the document]
Comment (pearu) [reply jax/_src/test_util.py:1521]: Yes, there are other uses where the arguments should be swapped to make the exception message clear.

I'll remove the comment.

File: tests/lax_scipy_special_functions_test.py [hunk 2]
```diff
@@ -317,6 +317,33 @@ def testGammaIncCBoundaryValues(self):
...
+    a_samples = [nan, -0.5, inf, 0, eps, 1, tiny][:-1]
+    b_samples = [nan, -0.5, inf, 0, eps, 1, tiny][:-1]
+    x_samples = [nan, -0.5, 0, 0.5, 1, 1.5]
+    if jtu.parse_version(scipy.__version__) >= (1, 16):
+      samples_slice = slice(None)
+    else:
+      # disable samples that contradict with scipy/scipy#22425
+      samples_slice = slice(None, len(a_samples) * len(b_samples) * 2)
+
+    args_maker = lambda: [
```
[the remark this answers is not in the document]
Comment (pearu) [reply tests/lax_scipy_special_functions_test.py]: Applied with a simplification to use `itertools.product`.

File: jax/_src/test_util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pearu) [reply jax/_src/test_util.py]: Fixed.

A followup PR: https://github.com/jax-ml/jax/pull/27255

File: tests/lax_scipy_special_functions_test.py [hunk 3]
```diff
@@ -317,6 +317,31 @@ def testGammaIncCBoundaryValues(self):
     self._CheckAgainstNumpy(lsp_special.gammaincc, osp_special.gammaincc, args_maker, rtol=rtol)
     self._CompileAndCheck(lsp_special.gammaincc, args_maker, rtol=rtol)
 
+  def testBetaIncBoundaryValues(self):
+    dtype = jax.numpy.zeros(0).dtype  # default float dtype.
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py:321]: More direct is `jax.dtypes.canonicalize_dtype(float)`

File: tests/lax_scipy_special_functions_test.py [hunk 4]
```diff
@@ -317,6 +317,31 @@ def testGammaIncCBoundaryValues(self):
...
+    x_samples = [nan, -0.5, 0, 0.5, 1, 1.5]
+    if jtu.parse_version(scipy.__version__) >= (1, 16):
+      samples_slice = slice(None)
+    else:
+      # disable samples that contradict with scipy/scipy#22425
+      samples_slice = slice(None, len(a_samples) * len(b_samples) * 2)
+
+    def args_maker():
+      return [np.array(lst[samples_slice]).astype(dtype)
+              for lst in zip(*itertools.product(a_samples, b_samples, x_samples))]
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py:340]: Maybe more clear to use `meshgrid`?
```suggestion
      return [x.ravel()[samples_slice] for x in np.meshgrid(
                  a_samples, b_samples, x_samples, indexing='ij')]
```
If `samples_slice` were no longer required, it would be simplified to this:
```python
      return [x.ravel() for x in np.meshgrid(a_samples, b_samples, x_samples)]
```
and since `betainc` accepts N-dimensional arrays, you could even simplify it to
```python
      return np.meshgrid(a_samples, b_samples, x_samples)
```

File: tests/lax_scipy_special_functions_test.py [hunk 5]
```diff
@@ -317,6 +317,31 @@ def testGammaIncCBoundaryValues(self):
...
+    # TODO(pearu): enable tiny samples when a fix to scipy/scipy#22682
+    # will be available
+    a_samples = [nan, -0.5, inf, 0, eps, 1, tiny][:-1]
+    b_samples = [nan, -0.5, inf, 0, eps, 1, tiny][:-1]
+    x_samples = [nan, -0.5, 0, 0.5, 1, 1.5]
+    if jtu.parse_version(scipy.__version__) >= (1, 16):
+      samples_slice = slice(None)
+    else:
+      # disable samples that contradict with scipy/scipy#22425
+      samples_slice = slice(None, len(a_samples) * len(b_samples) * 2)
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py:336]: I find it very hard to understand the effect of this. From running the code, it looks like maybe we could get rid of `samples_slice` and instead say `a_samples = a_samples[:2]`: is that accurate?

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_scipy_special_functions_test.py (hunk 3 above)
Comment (pearu) [reply tests/lax_scipy_special_functions_test.py]: Applied, thanks!

File: tests/lax_scipy_special_functions_test.py (hunk 4 above)
Comment (pearu) [reply tests/lax_scipy_special_functions_test.py]: Now using meshgrid.

File: tests/lax_scipy_special_functions_test.py (hunk 5 above)
Comment (pearu) [reply tests/lax_scipy_special_functions_test.py]: Specifying samples for different scipy versions is now simplified.

**Review by jakevdp — APPROVED** [review]
Looks good!

**Comment by jakevdp** [comment]
It looks like the test is failing with SciPy v1.11.1

**Review by jakevdp — APPROVED** [review]
